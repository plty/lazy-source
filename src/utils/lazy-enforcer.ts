import Thunk from '../interpreter/Thunk'
import { Value } from '../types'

export const unravel = (value: Thunk): Value => {
  // Used to check if there are any cyclic structures
  const unravelThunk = (v: Value | Thunk): Value => {
    if (v instanceof Thunk) {
      return v.value()
    } else if (Array.isArray(v)) {
      return v.map(x => unravelThunk(x))
    } else {
      return v
    }
  }

  return unravelThunk(value)
}