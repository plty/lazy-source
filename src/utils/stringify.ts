import { MAX_LIST_DISPLAY_LENGTH } from '../constants'
import Closure from '../interpreter/closure'
import { Value } from '../types'
import Thunk from '../interpreter/Thunk'

function makeIndent(indent: number | string): string {
  if (typeof indent === 'number') {
    if (indent > 10) {
      indent = 10
    }
    return ' '.repeat(indent)
  } else {
    if (indent.length > 10) {
      indent = indent.substring(0, 10)
    }
    return indent
  }
}

function indentify(indent: string, s: string): string {
  return s
    .split('\n')
    .map(v => indent + v)
    .join('\n')
}

export const stringify = function*(
  value: Thunk,
  indent: number | string = 2,
  splitlineThreshold = 80
): IterableIterator<Value> {
  // Used to check if there are any cyclic structures
  const ancestors = new Set()

  // Precompute useful strings
  const indentString = makeIndent(indent)
  const arrPrefix = '[' + indentString.substring(1)
  const objPrefix = '{' + indentString.substring(1)
  const arrSuffix = indentString.substring(0, indentString.length - 1) + ']'
  const objSuffix = indentString.substring(0, indentString.length - 1) + '}'

  // Util functions

  // Determines if the array/object containing these (stringified) values
  // should be multiline and indented or oneline
  const shouldMultiline = (valueStrs: string[]) =>
    indentString !== '' &&
    (valueStrs.join(', ').length > splitlineThreshold || valueStrs.some(s => s.includes('\n')))

  // Stringify functions
  // The real one is stringifyValue

  const stringifyArray = function*(xs: Value[], indentLevel: number) {
    ancestors.add(xs)
    const valueStrs = yield* concatGens(xs.map(x => () => stringifyValue(x, 0)))
    ancestors.delete(xs)

    if (shouldMultiline(valueStrs)) {
      if (xs.length === 2) {
        // It's (probably) a source list
        // Don't increase indent on second element
        // so long lists don't look like crap
        return `${arrPrefix}${indentify(
          indentString.repeat(indentLevel + 1),
          valueStrs[0]
        ).substring(indentString.length)},
${indentify(indentString.repeat(indentLevel), valueStrs[1])}${arrSuffix}`
      } else {
        // A regular array,
        // indent second element onwards to match with first element
        return `${arrPrefix}${indentify(
          indentString.repeat(indentLevel + 1),
          valueStrs.join(',\n')
        ).substring(indentString.length)}${arrSuffix}`
      }
    } else {
      return `[${valueStrs.join(', ')}]`
    }
  }

  const stringifyObject = function*(obj: object, indentLevel: number) {
    ancestors.add(obj)
    const keys: [string] = yield* concatGens(
      Object.entries(obj).map(entry => () => stringifyValue(Thunk.from(entry[0])), 0)
    )
    const vals: [string] = yield* concatGens(
      Object.entries(obj).map(entry => () => stringifyValue(Thunk.from(entry[1])), 0)
    )
    const valueStrs = zip(keys, vals).map(([key, val]) => {
      if (val.includes('\n')) {
        return key + ':\n' + indentify(indentString, val)
      } else {
        return key + ': ' + val
      }
    })
    ancestors.delete(obj)

    if (shouldMultiline(valueStrs)) {
      return `${objPrefix}${indentify(
        indentString.repeat(indentLevel + 1),
        valueStrs.join(',\n')
      ).substring(indentString.length)}${objSuffix}`
    } else {
      return `{${valueStrs.join(', ')}}`
    }
  }

  const stringifyValue = function*(thunk: Thunk, indentLevel: number = 0): IterableIterator<Value> {
    const v = yield* thunk.evaluate()
    if (v === null) {
      return 'null'
    } else if (v === undefined) {
      return 'undefined'
    } else if (ancestors.has(v)) {
      return '...<circular>'
    } else if (v instanceof Closure) {
      return v.toString()
    } else if (typeof v === 'string') {
      return JSON.stringify(v)
    } else if (typeof v !== 'object') {
      return v.toString()
    } else if (ancestors.size > MAX_LIST_DISPLAY_LENGTH) {
      return '...<truncated>'
    } else if (Array.isArray(v)) {
      return yield* stringifyArray(v, indentLevel)
    } else {
      return yield* stringifyObject(v, indentLevel)
    }
  }

  return yield* stringifyValue(value, 0)
}

function* concatGens(gens: (() => IterableIterator<Value>)[]): IterableIterator<Value> {
  return yield* gens.reduce(
    (acc, cur) => {
      return function*() {
        const rest = yield* acc()
        const v = yield* cur()
        return [...rest, v]
      }
    },
    function*() {
      return []
    }
  )()
}

function zip<T, U>(xs: T[], ys: U[]): [T, U][] {
  return xs.map((x, i) => [x, ys[i]])
}
