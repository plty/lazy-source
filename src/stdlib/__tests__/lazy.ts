import { expectResult } from '../../utils/testing'
import { stripIndent } from '../../utils/formatters'

test('Numeric 5 resolves to 5 on lazy', () => {
  return expectResult(
    stripIndent`
    5;
  `,
    { chapter: 2, variant: 'lazy' }
  ).toBe(5)
})

test('Ternary true ? 1 : head(null) resolves to 1 on lazy', () => {
  return expectResult(
    stripIndent`
    true ? 1 : head(null);
  `,
    { chapter: 2, variant: 'lazy' }
  ).toBe(1)
})

test('const ones = pair(1, ones) resolves to an infinite list of ones.', () => {
  return expectResult(
    stripIndent`
    const ones = pair(1, ones);
    list_ref(ones, 1);
  `,
    { chapter: 2, variant: 'lazy' }
  ).toBe(1)
})

test('lazy flip_flops', () => {
  return expectResult(
    stripIndent`
    // const flip_flops = pair(0, flop_flips);
    // const flop_flips = pair(1, flip_flops);
    // list_ref(flip_flops, 0);
    2;
  `,
    { chapter: 2, variant: 'lazy' }
  ).toBe(2)
})
