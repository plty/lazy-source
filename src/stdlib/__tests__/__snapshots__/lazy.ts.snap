// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Numeric 5 resolves to 5 on lazy: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "5;",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": 5,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Ternary true ? 1 : head(null) resolves to 1 on lazy: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "true ? 1 : head(null);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": 1,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`const ones = pair(1, ones) resolves to an infinite list of ones.: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "const ones = pair(1, ones);
list_ref(ones, 1);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": 1,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`lazy flip_flops: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "// const flip_flops = pair(0, flop_flips);
// const flop_flips = pair(1, flip_flops);
// list_ref(flip_flops, 0);
2;",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": 2,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;
